pkg/scaling/scale_handler.go

```diff
@@ -1,77 +1,96 @@
+	GetScalers(scalableObject interface{}) ([]scalers.Scaler, error)
+}
+
+type scaleHandler struct {
+	client            client.Client
+	logger            logr.Logger
+	scaleLoopContexts *sync.Map
+	scaleExecutor     executor.ScaleExecutor
 }
 
 const (
 	// Default polling interval for a ScaledObject triggers if no pollingInterval is defined.
 	defaultPollingInterval = 30
-	// Default cooldown period for a ScaleTarget if no cooldownPeriod is defined on the scaledObject
-	defaultCooldownPeriod = 5 * 60 // 5 minutes
 )
 
 // NewScaleHandler creates a ScaleHandler object
-func NewScaleHandler(client client.Client, scaleClient *scale.ScalesGetter, reconcilerScheme *runtime.Scheme) *ScaleHandler {
-	handler := &ScaleHandler{
-		client:           client,
-		scaleClient:      scaleClient,
-		logger:           logf.Log.WithName("scalehandler"),
-		reconcilerScheme: reconcilerScheme,
+func NewScaleHandler(client client.Client, scaleClient *scale.ScalesGetter, reconcilerScheme *runtime.Scheme) ScaleHandler {
+	return &scaleHandler{
+		client:            client,
+		logger:            logf.Log.WithName("scalehandler"),
+		scaleLoopContexts: &sync.Map{},
+		scaleExecutor:     executor.NewScaleExecutor(client, scaleClient, reconcilerScheme),
 	}
-	return handler
 }
 
-func (h *ScaleHandler) updateScaledObjectStatus(scaledObject *kedav1alpha1.ScaledObject) error {
-	err := h.client.Status().Update(context.TODO(), scaledObject)
+func (h *scaleHandler) GetScalers(scalableObject interface{}) ([]scalers.Scaler, error) {
+	withTriggers, err := asDuckWithTriggers(scalableObject)
 	if err != nil {
-		if errors.IsConflict(err) {
-			// ScaledObject's metadata that are not necessary to restart the ScaleLoop were updated (eg. labels)
-			// we should try to fetch the scaledObject again and process the update once again
-			h.logger.V(1).Info("Trying to fetch updated version of ScaledObject to properly update it's Status")
-			updatedScaledObject := &kedav1alpha1.ScaledObject{}
-			err2 := h.client.Get(context.TODO(), types.NamespacedName{Name: scaledObject.Name, Namespace: scaledObject.Namespace}, updatedScaledObject)
-			if err2 != nil {
-				h.logger.Error(err2, "Error getting updated version of ScaledObject before updating it's Status")
-			} else {
-				scaledObject = updatedScaledObject
-				if h.client.Status().Update(context.TODO(), scaledObject) == nil {
-					h.logger.V(1).Info("ScaledObject's Status was properly updated on re-fetched ScaledObject")
-					return nil
-				}
-			}
-		}
-		// we got another type of error
-		h.logger.Error(err, "Error updating scaledObject status")
-		return err
+		return nil, err
+	}
+
+	withPods, containerName, err := h.resolvePods(scalableObject)
+	if err != nil {
+		return nil, err
+	}
+
+	return h.getScalers(withTriggers, withPods, containerName)
+}
+
+func asDuckWithTriggers(scalableObject interface{}) (*kedav1alpha1.WithTriggers, error) {
+	switch scalableObject.(type) {
```

Comment:
I hthink we could avoid the object->unstructure->duck-type conversion. This would be needed if we don't know the specific type (and might be probably used if we define some generic Keda Duck type, which could be used next to ScaledObject and ScaledJob).

But since we know both types, the conversion could be done like this:
```go
func asDuckWithTriggers(scalableObject interface{}) (*kedav1alpha1.WithTriggers, error) {
	withTriggers := &kedav1alpha1.WithTriggers{}
	switch obj := scalableObject.(type) {
	case *kedav1alpha1.ScaledObject:
		withTriggers = &kedav1alpha1.WithTriggers{
			Spec: kedav1alpha1.WithTriggersSpec{
				PollingInterval: obj.Spec.PollingInterval,
				Triggers: obj.Spec.Triggers,
			},
		}
	case *kedav1alpha1.ScaledJob:
		withTriggers = &kedav1alpha1.WithTriggers{
			Spec: kedav1alpha1.WithTriggersSpec{
				PollingInterval: obj.Spec.PollingInterval,
				Triggers: obj.Spec.Triggers,
			},
		}
         default:
                 // here could be the conversion from unknown Duck type potentially in the future
		return nil, fmt.Errorf("unknown scalable object type %v", scalableObject)
	}
	return withTriggers, nil
}
```

Reply:
That's indeed much better, and makes sense. I was so fixated on using `duck.FromUnstructured` for some reason. Thanks!!